cpp/comms/std/src/ucp_helper.h

```diff
@@ -0,0 +1,141 @@
+/*
+ * Copyright (c) 2019, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <ucp/api/ucp.h>
+#include <ucp/api/ucp_def.h>
+
+#include <utils.h>
+static const ucp_tag_t default_tag_mask = -1;
+
+/**
+ * @brief Continues progress on worker until a message request has completed
+ */
+static void wait(ucp_worker_h ucp_worker, struct ucx_context *context) {
+  while (context->completed == 0) {
+    ucp_worker_progress(ucp_worker);
+  }
+}
+
+/**
+ * @brief callback for flushing worker
+ */
+static void flush_callback(void *request, ucs_status_t status) {}
+
+/**
+ * @brief Flush the send buffer on the given endpoint for the given worker
+ */
+static ucs_status_t flush_ep(ucp_worker_h worker, ucp_ep_h ep) {
+  void *request;
+
+  request = ucp_ep_flush_nb(ep, 0, flush_callback);
+  if (request == NULL) {
+    return UCS_OK;
+  } else if (UCS_PTR_IS_ERR(request)) {
+    return UCS_PTR_STATUS(request);
+  } else {
+    ucs_status_t status;
+    do {
+      ucp_worker_progress(worker);
+      status = ucp_request_check_status(request);
+    } while (status == UCS_INPROGRESS);
+    ucp_request_release(request);
+    return status;
+  }
+}
+
+/**
+ * @brief Asynchronous send callback sets request to completed
+ */
+static void send_handle(void *request, ucs_status_t status) {
+  struct ucx_context *context = (struct ucx_context *)request;
+  context->completed = 1;
+}
+
+/**
+ * @brief Asynchronous recv callback sets request to completed
+ */
+static void recv_handle(void *request, ucs_status_t status,
+                        ucp_tag_recv_info_t *info) {
+  struct ucx_context *context = (struct ucx_context *)request;
+  context->completed = 1;
+}
+
+/**
+ * @brief Asynchronously send data to the given endpoint using the given tag
+ */
+struct ucx_context *ucp_isend(ucp_ep_h ep_ptr, const void *buf, int size,
+                              int tag) {
+  struct ucx_context *ucp_request = 0;
+  ucp_tag_t ucp_tag = (ucp_tag_t)tag;
+
+  ucp_request = (struct ucx_context *)ucp_tag_send_nb(
+    ep_ptr, buf, size, ucp_dt_make_contig(1), ucp_tag, send_handle);
+
+  /**
+   * On error, close endpoint, set request to completed and close
+   * @TODO: Should the request be closed here as well? How should we handle this?
+   * @TODO: Should the Python layer be attempting to reconnect endpoints when they are closed?
+   */
+  if (UCS_PTR_IS_ERR(ucp_request)) {
+    ucp_ep_close_nb(ep_ptr, UCP_EP_CLOSE_MODE_FLUSH);
```

Comment:
Should closing the ep fall under the scope of `isend`?

Reply:
What I mean to say is that send could fail for other reasons and not necessarily a faulty endpoint. For example, it could be a bad buffer. What are your thoughts?

Reply:
> Should closing the ep fall under the scope of isend?

This was another example I believe I took from `hello_world.c` which was unclear to me, and the reason for the TODOs. The call to `ASSERT` below this line will throw an exception, which I believe will stop all future operations. 

Do you think it would be better here to just print a message to stderr and try to continue or should we introspect the error to see the best way to handle it? It was unclear to me from the `hello_world.c` the best way to handle errors.

Reply:
Some of the error types [here](https://github.com/openucx/ucx/blob/master/src/ucs/type/status.h#L45) are applicable. Could be better of to just print and try to continue for now? I'm not really sure about it though.

Reply:
I have opened #845 to capture better handling / recovery of UCx errors.  This might be a post 0.9 improvement. For now, I’ll use the standard ASSERT created for cuML.